terms/fast_fourier_transforms.md

```diff
@@ -1,3 +1,152 @@
+
+The Fourier Transform we discuss is a discrete Fourier Transform over the [finite field](./prime_or_finite_fields.md), rather than the
+continuous Fourier Transform over the *real or complex number* above. And, instead of focusing on the conversion between
+"frequency domain" and "time domain", we primarily talk about two operations: **multi-point polynomial evaluation** and its
+inverse, **polynomial interpolation**.
+
+First, **multi-point polynomial evaluation** involves evaluating a degree $\lt N$ polynomial at $N$ different points, and
+**polynomial interpolation** is recovering the polynomial using the given evaluation of its at $N$ different points.
+
+For the first operation, you all can separately evaluate the polynomial at each point just by iterating over all the coefficients.
+Thus, the total runtime is $O(N^2)$ (considering $N$ coefficients and $N$ points).
+For the second operation, you first need to read about [Lagrange interpolation](./lagrange_interpolation.md) to understand how it
+works and what Lagrange basis polynomials are:
+$$L_i(x) = \prod_{j=0, j \neq i}^{n} \frac{x - x_j}{x_i - x_j}$$
+Compute this function takes $O(N^2)$ time (iterate $j$ take $O(N)$ times, and we need to compute $N$ functions).
+
+Is this slow? Yes, and the Fast Fourier Transform below will significantly speed up both operations.
+
+## Fast Fourier Transform
+
+To use this algorithm, you need to use finite field, and the domain must be a [multiplicative subgroup](./prime_or_finite_fields) of
+the field. Furthermore, the multiplicative subgroup must have size $2^n$ in order to create a low-runtime recursive algorithm.
+
+**High level idea:**
+FFT works like [merge-sort](https://www.geeksforgeeks.org/merge-sort/) . It takes a problem of size $N$, break it up into two problems
+of size $\dfrac{N}{2}$, and do $O(N)$ work to combine two smaller solutions into the bigger one. So, the total run time is $O(NlogN)$.
+
+### Multi-point evaluation
+
+Suppose that we have an FFT where the given domain is the powers of $w$ in some field, where $w^{2^k} = 1$ ($w$ is the domain
+"generator").
+
+For example, consider the finite field $F_{17}$, a polynomial: $f(x) = 1 + 13x + 3x^2 + 3x^3$ (we will write it as \[1, 13, 3, 3\]).
+We want to evaluate this polynomial at each point in the domain, i.e., at the multiplicative subgroup of $4$:
+$\lbrace 1, 4, 16, 13 \rbrace$.
+
+First, we break up $f(x)$ into two parts, which we will call $evens$ and $odds$:
+
+- $evens$: \[1, 3\] or  $evens(x) = 1 + 3x$
+- $odds$: \[13, 3\] or $odds(x) = 13 + 3x$
+
+So, we have:  
+
+- $f(x) = evens(x^2) + x \cdot odds(x^2)$
+- $f(-x) = evens(x^2) - x \cdot odds(x^2)$
+
+Here, both $evens$ and $odds$ are polynomials half the size of $f(x)$ and the set of possible values of $x^2$ is only half the size of
+the original domain: \[1, 4, 16, 13\] $\to$ \[1, 16\]. The reason is that:
+
+- $1^2 = 1$
+- $4^2 = 16$
+- $16^2 = 256 \equiv 1\mod 17$
+- $13^2 = 169 \equiv 16 \mod 17$
```

Comment:
for more concrete, $1 \equiv \omega^{|G|}$, so $(\omega^k)^2 \equiv (\omega^{k+|G|/2})^2$